pkg/multicast/multicast.go

```diff
@@ -0,0 +1,152 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package multicast
+
+import (
+	"net"
+
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"golang.org/x/net/ipv6"
+)
+
+var (
+	// v6Socket is the udp socket used to join/leave a multicast group
+	v6Socket net.PacketConn
+
+	mutex lock.Mutex
+
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "multicast")
+)
+
+// Pre-Defined Multicast Addresses
+// as defined in https://tools.ietf.org/html/rfc4291#section-2.7.1
+
+var (
+	// AllNodesIfcLocalMaddr is the multicast address that identifies the group of
+	// all IPv6 nodes, within scope 1 (interface-local)
+	AllNodesIfcLocalMaddr net.IP = net.ParseIP("ff01::1")
+
+	// AllNodesLinkLocalMaddr is the multicast address that identifies the group of
+	// all IPv6 nodes, within scope 2 (link-local)
+	AllNodesLinkLocalMaddr net.IP = net.ParseIP("ff02::1")
+
+	// AllRoutersIfcLocalMaddr is the multicast address that identifies the group of
+	// all IPv6 routers, within scope 1 (interface-local)
+	AllRoutersIfcLocalMaddr net.IP = net.ParseIP("ff01::2")
+
+	// AllRoutersLinkLocalMaddr is the multicast address that identifies the group of
+	// all IPv6 routers, within scope 2 (link-local)
+	AllRoutersLinkLocalMaddr net.IP = net.ParseIP("ff02::2")
+
+	// AllRoutersSiteLocalMaddr is the multicast address that identifies the group of
+	// all IPv6 routers, within scope 5 (site-local)
+	AllRoutersSiteLocalMaddr net.IP = net.ParseIP("ff05::2")
+
+	// SolicitedNodeMaddrPrefix is the prefix of the multicast address that is used
+	// as part of NDP
+	SolicitedNodeMaddrPrefix net.IP = net.ParseIP("ff02::1:ff00:0")
+)
+
+func initSocket() error {
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	if v6Socket != nil {
+		return nil
+	}
+
+	c, err := net.ListenPacket("udp6", "[::]:22334")
```

Comment:
Is there anything special about this port 22334? Just wondering from a high-level whether this listen path could conflict with anything and hence should be configurable via CLI.

Reply:
It doesn't matter which port to use. Should I use a random port instead?

Reply:
Do we need to specify the port at all? Seems like if we leave it empty, the go libraries will pick one for us. That way it'll resolve any potential conflicts with other things in the system.

The only caveat I can think is if we need to ensure it's a specific port to allow users to poke holes in their firewalls for this traffic (whether the Cilium hostfw or other firewalls in their network). If this is the case, the current number is fine but:
(1) It'd be good to plumb this up to a configuration option so if it's a problem the user can swap the port
(2) We should document the default port we use here: https://docs.cilium.io/en/v1.8/install/system_requirements/#firewall-rules